File: Problems & Solutions/144. Binary Tree Preorder Traversal/09 January 2022/Approach 3 - Morris Traversal/Binary Tree Preorder Traversal.cpp
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> answer;
        TreeNode* curr = root;
        TreeNode* last;
        
        while (curr != nullptr) {
            if (curr->left == nullptr) {
                answer.push_back(curr->val);
                curr = curr->right;
            } else {
                last = curr->left;
                while (last->right != nullptr && last->right != curr) {
                    last = last->right;
                }
                
                if (last->right == nullptr) {
                    answer.push_back(curr->val);
                    last->right = curr;
                    curr = curr->left;
                } else {
                    last->right = nullptr;
                    curr = curr->right;
                }
            }
        }

        
        return answer;
    }
};

File: Problems & Solutions/144. Binary Tree Preorder Traversal/09 January 2022/Approach 3 - Morris Traversal/Binary Tree Preorder Traversal.java
class Solution {    
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> answer = new ArrayList<>();
        TreeNode curr = root, last;
        
        while (curr != null) {
            if(curr.left == null) {
                answer.add(curr.val);
                curr = curr.right;
            } else {
                last = curr.left;
                while (last.right != null && last.right != curr) {
                    last = last.right;
                }

                if (last.right == null) {
                    answer.add(curr.val);
                    last.right = curr;
                    curr = curr.left;
                } else {
                    last.right = null;
                    curr = curr.right;
                }
            }
        }
        
        return answer;
    }
}